### **Machine Learning Project - Group AR**

In [89]:
# 1) Core imports
from pathlib import Path
import warnings

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

warnings.filterwarnings("ignore")


# Load dataset from GitHub commit permalink (online, reproducible)
DATA_URL = "https://raw.githubusercontent.com/Behnia02/Group_AR/631a8d5f9337e2a4b9351ffb13ca583d28459dc1/HR-Employee-Attrition.csv"

df = pd.read_csv(DATA_URL)

print(f"Dataset loaded online from: {DATA_URL}")
print(f"Shape: {df.shape[0]} rows x {df.shape[1]} columns")
display(df.head())

Dataset loaded online from: https://raw.githubusercontent.com/Behnia02/Group_AR/631a8d5f9337e2a4b9351ffb13ca583d28459dc1/HR-Employee-Attrition.csv
Shape: 1470 rows x 35 columns


Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0        19479                   8      Y      Yes                 11   
1        24907                   1      Y       No                 23   
2         2396                   6      Y      Yes                 15   
3        23159                   1      Y      Yes                 11   
4        16632                   9      Y       No                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

#### **1. Data Cleaning**

In [90]:
# Build summary table
summary_table = pd.DataFrame({
    "Total Values": df.count(),
    "Null Values": df.isna().sum(),
    "Unique Values": df.nunique()})

# Display the summary table
summary_table.head(len(summary_table))

Total Values  Null Values  Unique Values
Age                               1470            0             43
Attrition                         1470            0              2
BusinessTravel                    1470            0              3
DailyRate                         1470            0            886
Department                        1470            0              3
DistanceFromHome                  1470            0             29
Education                         1470            0              5
EducationField                    1470            0              6
EmployeeCount                     1470            0              1
EmployeeNumber                    1470            0           1470
EnvironmentSatisfaction           1470            0              4
Gender                            1470            0              2
HourlyRate                        1470            0             71
JobInvolvement                    1470            0              4
JobLevel                          1470            0              5
JobRole                           1470            0              9
JobSatisfaction                   1470            0              4
MaritalStatus                     1470            0              3
MonthlyIncome                     1470            0           1349
MonthlyRate                       1470            0           1427
NumCompaniesWorked                1470            0             10
Over18                            1470            0              1
OverTime                          1470            0              2
PercentSalaryHike                 1470            0             15
PerformanceRating                 1470            0              2
RelationshipSatisfaction          1470            0              4
StandardHours                     1470            0              1
StockOptionLevel                  1470            0              4
TotalWorkingYears                 1470            0             40
TrainingTimesLastYear             1470            0              7
WorkLifeBalance                   1470            0              4
YearsAtCompany                    1470            0             37
YearsInCurrentRole                1470            0             19
YearsSinceLastPromotion           1470            0             16
YearsWithCurrManager              1470            0             18

From this table, we can conclude that there is no missing data in the dataset, since all variables have complete records across the 1470 observations.

**Handling redundant data**

In [91]:
#Look for duplicates
print("Duplicates:", df.duplicated().sum())

Duplicates: 0


In [92]:
# Redundant columns - variables with only one unique value

one_unique_vars = summary_table[summary_table["Unique Values"] == 1].index.tolist()
print(one_unique_vars)


['EmployeeCount', 'Over18', 'StandardHours']


In [93]:
# Drop the columns with just one unique value and EmployeeNumber (ID variable)
df = df.drop(columns=["EmployeeCount", "Over18", "StandardHours", "EmployeeNumber"], errors="ignore")

**Convenience encoding**

In [94]:
#Printing out binary variables
import pandas as pd

# --- Identify binary variables (include Over18, exclude PerformanceRating) ---
binary_vars = df[[col for col in df.columns
                  if (df[col].nunique() == 2 and col != "PerformanceRating")
                  ]]

# --- Pretty display of the first rows ---
binary_vars.head().style.set_properties(**{
    'border': '1px solid black',
    'padding': '4px'
}).set_table_styles([{
    'selector': 'th',
    'props': [('background-color', '#d3d3d3'),
              ('color', 'black'),
              ('font-weight', 'bold'),
              ('text-align', 'center')]
}])

**Transforming Binary Variables**

In [95]:


#Turning binary variables to numeric
# --- Copy dataset to avoid modifying original ---
df = df.copy()

# --- Explicit binary mappings ---
df["Attrition_Flag"] = df["Attrition"].map({"Yes": 1, "No": 0})
df["OverTime_Flag"]  = df["OverTime"].map({"Yes": 1, "No": 0})
df["Gender_Flag"]    = df["Gender"].map({"Male": 1, "Female": 0})


# --- Styled output preview ---
df_preview = df[[
    "Attrition", "Attrition_Flag",
    "OverTime", "OverTime_Flag",
    "Gender", "Gender_Flag",

]].head()

df_preview.style.set_properties(**{
    'border': '1px solid black',
    'padding': '4px'
}).set_table_styles([{
    'selector': 'th',
    'props': [('background-color', '#d3d3d3'),
              ('color', 'black'),
              ('font-weight', 'bold'),
              ('text-align', 'center')]
}])


In [96]:
#Drop original categorical columns after encoding
df = df.drop(columns=["Attrition", "OverTime", "Gender"])

**Missing Value Check**

In [97]:

#Checking for missing values
if df.isnull().values.any():
    print("There are missing values in the dataset.")
else:
    print("No missing values found")


No missing values found


In [98]:
#Sanity check
pd.set_option('display.max_columns', None)
df.head()

Age     BusinessTravel  DailyRate              Department  \
0   41      Travel_Rarely       1102                   Sales   
1   49  Travel_Frequently        279  Research & Development   
2   37      Travel_Rarely       1373  Research & Development   
3   33  Travel_Frequently       1392  Research & Development   
4   27      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EnvironmentSatisfaction  \
0                 1          2  Life Sciences                        2   
1                 8          1  Life Sciences                        3   
2                 2          2          Other                        4   
3                 3          4  Life Sciences                        4   
4                 2          1        Medical                        1   

   HourlyRate  JobInvolvement  JobLevel                JobRole  \
0          94               3         2        Sales Executive   
1          61               2         2     Research Scientist   
2          92               2         1  Laboratory Technician   
3          56               3         1     Research Scientist   
4          40               3         1  Laboratory Technician   

   JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  \
0                4        Single           5993        19479   
1                2       Married           5130        24907   
2                3        Single           2090         2396   
3                3       Married           2909        23159   
4                2       Married           3468        16632   

   NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
0                   8                 11                  3   
1                   1                 23                  4   
2                   6                 15                  3   
3                   1                 11                  3   
4                   9                 12                  3   

   RelationshipSatisfaction  StockOptionLevel  TotalWorkingYears  \
0                         1                 0                  8   
1                         4                 1                 10   
2                         2                 0                  7   
3                         3                 0                  8   
4                         4                 1                  6   

   TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                      0                1               6                   4   
1                      3                3              10                   7   
2                      3                3               0                   0   
3                      3                3               8                   7   
4                      3                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  Attrition_Flag  \
0                        0                     5               1   
1                        1                     7               0   
2                        0                     0               1   
3                        3                     0               0   
4                        2                     2               0   

   OverTime_Flag  Gender_Flag  
0              1            0  
1              0            1  
2              1            1  
3              1            0  
4              0            1

**Classification of the variables**

*Classification by type*

In [99]:
# TARGET
target = "Attrition_Flag"

# BINARY VARIABLES (Mapped)
binary_vars = [
    "Attrition_Flag",   # Target
    "OverTime_Flag",
    "Gender_Flag"
]

# CATEGORICAL VARIABLES (Nominal)
categorical_vars = [
    "BusinessTravel",
    "Department",
    "EducationField",
    "JobRole",
    "MaritalStatus"
]

# ORDINAL / DISCRETE NUMERIC
ordinal_vars = [
    "Education",
    "EnvironmentSatisfaction",
    "JobInvolvement",
    "JobLevel",
    "JobSatisfaction",
    "PerformanceRating",
    "RelationshipSatisfaction",
    "StockOptionLevel",
    "TrainingTimesLastYear",
    "WorkLifeBalance"
]

# CONTINUOUS NUMERIC
continuous_vars = [
    "Age",
    "DailyRate",
    "DistanceFromHome",
    "HourlyRate",
    "MonthlyIncome",
    "MonthlyRate",
    "NumCompaniesWorked",
    "PercentSalaryHike",
    "TotalWorkingYears",
    "YearsAtCompany",
    "YearsInCurrentRole",
    "YearsSinceLastPromotion",
    "YearsWithCurrManager"
]

# FINAL FEATURE LIST (FOR MODELING)

feature_vars = (
    binary_vars[1:] +   # exclude target
    categorical_vars +
    ordinal_vars +
    continuous_vars
)

*Classification by clusters*

In [100]:
# TARGET
target = "Attrition_Flag"


# Demographics
demographic_vars = [
    "Age",
    "Gender_Flag",
    "MaritalStatus",
    "Education",
    "EducationField",
    "DistanceFromHome"
]


# Compensation & Financial Incentives
compensation_vars = [
    "MonthlyIncome",
    "DailyRate",
    "HourlyRate",
    "MonthlyRate",
    "PercentSalaryHike",
    "StockOptionLevel",
    "JobLevel"
]


# Job Structure & Role Characteristics
job_structure_vars = [
    "Department",
    "JobRole",
    "BusinessTravel",
    "OverTime_Flag",
    "JobInvolvement"
]


# Career Progression & Tenure
career_progression_vars = [
    "TotalWorkingYears",
    "YearsAtCompany",
    "YearsInCurrentRole",
    "YearsSinceLastPromotion",
    "YearsWithCurrManager",
    "NumCompaniesWorked",
    "TrainingTimesLastYear"
]


# Workplace Satisfaction & Environment
satisfaction_vars = [
    "JobSatisfaction",
    "EnvironmentSatisfaction",
    "RelationshipSatisfaction",
    "WorkLifeBalance",
    "PerformanceRating"
]



In [101]:
### TO DELETE


#Sanity Check: Check for duplicates in feature list
len(feature_vars), len(set(feature_vars))


(30, 30)

**Metatable of the included variables**

In [102]:
# Metadata table based on current df only

def infer_var_type(series):
    if pd.api.types.is_numeric_dtype(series):
        if series.nunique() == 2:
            return "Binary"
        if series.nunique() <= 10:
            return "Ordinal"
        return "Numerical"
    return "Nominal"

# Optional short descriptions for common columns (only used if present in df)
descriptions = {
    "Age": "Employee age in years",
    "Attrition_Flag": "Binary target: attrition",
    "BusinessTravel": "Frequency of business travel",
    "DailyRate": "Daily rate of employee",
    "Department": "Department of the employee",
    "DistanceFromHome": "Distance from home to office",
    "Education": "Education level (1-5)",
    "EducationField": "Field of education",
    "EnvironmentSatisfaction": "Satisfaction with work environment (1-4)",
    "Gender_Flag": "Binary gender flag",
    "HourlyRate": "Hourly pay rate",
    "JobInvolvement": "Job involvement (1-4)",
    "JobLevel": "Job hierarchy level",
    "JobRole": "Role or position in company",
    "JobSatisfaction": "Job satisfaction (1-4)",
    "MaritalStatus": "Marital status of employee",
    "MonthlyIncome": "Monthly salary",
    "MonthlyRate": "Monthly rate for payroll purposes",
    "NumCompaniesWorked": "Number of previous companies worked",
    "OverTime_Flag": "Binary overtime flag",
    "PercentSalaryHike": "Percentage salary increase",
    "PerformanceRating": "Performance rating (1-4)",
    "RelationshipSatisfaction": "Relationship satisfaction (1-4)",
    "StockOptionLevel": "Stock option level",
    "TotalWorkingYears": "Total years of work experience",
    "TrainingTimesLastYear": "Number of trainings attended last year",
    "WorkLifeBalance": "Work-life balance rating (1-4)",
    "YearsAtCompany": "Years spent at current company",
    "YearsInCurrentRole": "Years spent in current role",
    "YearsSinceLastPromotion": "Years since last promotion",
    "YearsWithCurrManager": "Years working with current manager"
}

# Build metadata table for all columns in df
metadata_df = pd.DataFrame([
    [col, str(df[col].dtype), infer_var_type(df[col]), descriptions.get(col, "-")]
    for col in df.columns
], columns=["Variable Name", "Data Type", "Variable Type", "Description"])

metadata_styled = (
    metadata_df.style
    .set_properties(**{'border': '1px solid black', 'padding': '4px'})
    .set_table_styles([{
        'selector': 'th',
        'props': [('background-color', '#d3d3d3'),
                  ('color', 'black'),
                  ('font-weight', 'bold'),
                  ('text-align', 'center')]
    }])
    .hide(axis="index")
)

metadata_styled

#### **2. EDA**

**Descriptive Statistics**

In [ ]:
# Descriptive statistics for numeric variables
num_cols = df.select_dtypes(include="number").columns.tolist()

numeric_summary = df[num_cols].describe().T
numeric_summary["missing"] = df[num_cols].isna().sum()
numeric_summary["missing_pct"] = (numeric_summary["missing"] / len(df) * 100).round(2)

numeric_summary

count          mean          std     min     25%  \
Age                       1470.0     36.923810     9.135373    18.0    30.0   
DailyRate                 1470.0    802.485714   403.509100   102.0   465.0   
DistanceFromHome          1470.0      9.192517     8.106864     1.0     2.0   
Education                 1470.0      2.912925     1.024165     1.0     2.0   
EnvironmentSatisfaction   1470.0      2.721769     1.093082     1.0     2.0   
HourlyRate                1470.0     65.891156    20.329428    30.0    48.0   
JobInvolvement            1470.0      2.729932     0.711561     1.0     2.0   
JobLevel                  1470.0      2.063946     1.106940     1.0     1.0   
JobSatisfaction           1470.0      2.728571     1.102846     1.0     2.0   
MonthlyIncome             1470.0   6502.931293  4707.956783  1009.0  2911.0   
MonthlyRate               1470.0  14313.103401  7117.786044  2094.0  8047.0   
NumCompaniesWorked        1470.0      2.693197     2.498009     0.0     1.0   
PercentSalaryHike         1470.0     15.209524     3.659938    11.0    12.0   
PerformanceRating         1470.0      3.153741     0.360824     3.0     3.0   
RelationshipSatisfaction  1470.0      2.712245     1.081209     1.0     2.0   
StockOptionLevel          1470.0      0.793878     0.852077     0.0     0.0   
TotalWorkingYears         1470.0     11.279592     7.780782     0.0     6.0   
TrainingTimesLastYear     1470.0      2.799320     1.289271     0.0     2.0   
WorkLifeBalance           1470.0      2.761224     0.706476     1.0     2.0   
YearsAtCompany            1470.0      7.008163     6.126525     0.0     3.0   
YearsInCurrentRole        1470.0      4.229252     3.623137     0.0     2.0   
YearsSinceLastPromotion   1470.0      2.187755     3.222430     0.0     0.0   
YearsWithCurrManager      1470.0      4.123129     3.568136     0.0     2.0   
Attrition_Flag            1470.0      0.161224     0.367863     0.0     0.0   
OverTime_Flag             1470.0      0.282993     0.450606     0.0     0.0   
Gender_Flag               1470.0      0.600000     0.490065     0.0     0.0   

                              50%       75%      max  missing  missing_pct  
Age                          36.0     43.00     60.0        0          0.0  
DailyRate                   802.0   1157.00   1499.0        0          0.0  
DistanceFromHome              7.0     14.00     29.0        0          0.0  
Education                     3.0      4.00      5.0        0          0.0  
EnvironmentSatisfaction       3.0      4.00      4.0        0          0.0  
HourlyRate                   66.0     83.75    100.0        0          0.0  
JobInvolvement                3.0      3.00      4.0        0          0.0  
JobLevel                      2.0      3.00      5.0        0          0.0  
JobSatisfaction               3.0      4.00      4.0        0          0.0  
MonthlyIncome              4919.0   8379.00  19999.0        0          0.0  
MonthlyRate               14235.5  20461.50  26999.0        0          0.0  
NumCompaniesWorked            2.0      4.00      9.0        0          0.0  
PercentSalaryHike            14.0     18.00     25.0        0          0.0  
PerformanceRating             3.0      3.00      4.0        0          0.0  
RelationshipSatisfaction      3.0      4.00      4.0        0          0.0  
StockOptionLevel              1.0      1.00      3.0        0          0.0  
TotalWorkingYears            10.0     15.00     40.0        0          0.0  
TrainingTimesLastYear         3.0      3.00      6.0        0          0.0  
WorkLifeBalance               3.0      3.00      4.0        0          0.0  
YearsAtCompany                5.0      9.00     40.0        0          0.0  
YearsInCurrentRole            3.0      7.00     18.0        0          0.0  
YearsSinceLastPromotion       1.0      3.00     15.0        0          0.0  
YearsWithCurrManager          3.0      7.00     17.0        0          0.0  
Attrition_Flag                0.0      0.00  

**Categorical distributions**

In [105]:
# Frequency tables for categorical variables
cat_cols = df.select_dtypes(exclude="number").columns.tolist()

cat_summary = []
for col in cat_cols:
    counts = df[col].value_counts(dropna=False)
    top = counts.head(10)
    for level, count in top.items():
        cat_summary.append({
            "Variable": col,
            "Level": str(level),
            "Count": int(count),
            "Percent": round(count / len(df) * 100, 2)
        })

cat_summary_df = pd.DataFrame(cat_summary)
cat_summary_df

Variable                      Level  Count  Percent
0   BusinessTravel              Travel_Rarely   1043    70.95
1   BusinessTravel          Travel_Frequently    277    18.84
2   BusinessTravel                 Non-Travel    150    10.20
3       Department     Research & Development    961    65.37
4       Department                      Sales    446    30.34
5       Department            Human Resources     63     4.29
6   EducationField              Life Sciences    606    41.22
7   EducationField                    Medical    464    31.56
8   EducationField                  Marketing    159    10.82
9   EducationField           Technical Degree    132     8.98
10  EducationField                      Other     82     5.58
11  EducationField            Human Resources     27     1.84
12         JobRole            Sales Executive    326    22.18
13         JobRole         Research Scientist    292    19.86
14         JobRole      Laboratory Technician    259    17.62
15         JobRole     Manufacturing Director    145     9.86
16         JobRole  Healthcare Representative    131     8.91
17         JobRole                    Manager    102     6.94
18         JobRole       Sales Representative     83     5.65
19         JobRole          Research Director     80     5.44
20         JobRole            Human Resources     52     3.54
21   MaritalStatus                    Married    673    45.78
22   MaritalStatus                     Single    470    31.97
23   MaritalStatus                   Divorced    327    22.24